lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+        )
+
+        for authority_provided_id in authority_provided_ids:
+            if authority_provided_id in self._unified_courses:
+                continue
+
+            first_grouping = self._db.scalars(
+                select(Grouping)
+                .filter_by(authority_provided_id=authority_provided_id)
+                .order_by(Grouping.updated.desc())
+            ).first()
+
+            if not first_grouping:
+                continue
+
+            if first_grouping.type == Grouping.Type.COURSE:
+                course_authority_provided_id = first_grouping.authority_provided_id
+            else:
+                assert first_grouping.parent.type == Grouping.Type.COURSE
+                course_authority_provided_id = (
+                    first_grouping.parent.authority_provided_id
+                )
+
+            # Find all the course groupings for course_authority_provided_id (across all application instances).
+            courses = self._db.scalars(
+                select(Course)
+                .filter_by(authority_provided_id=course_authority_provided_id)
+                .order_by(Course.updated.desc())
+            ).all()
+
+            # Find all the sub-groupings of all the courses, across application instances.
+            sub_groups = self._db.scalars(
+                select(Grouping).filter(
+                    Grouping.parent_id.in_([course.id for course in courses])
+                )
+            ).all()
+
+            # Find all known instructors in the course.
+            instructors = [
+                unified_user
+                for unified_user in self.unified_users.values()
+                if is_instructor(unified_user, courses)
+            ]
+
+            # The authority_provided_ids of all the course groupings and sub-groupings for this course.
+            course_authority_provided_ids = set(
+                grouping.authority_provided_id for grouping in courses + sub_groups
+            )
+
+            # All the learner annotations in this course.
+            learner_annotations = [
+                annotation
+                for annotation in self._annotations
+                if annotation["group"]["authority_provided_id"]
+                in course_authority_provided_ids
+                and self.unified_users[annotation["author"]["userid"]]
+                not in instructors
+            ]
+
+            unified_course = UnifiedCourse(
+                authority_provided_id=course_authority_provided_id,
+                title=courses[0].lms_name,
+                instructors=tuple(instructors),
+                learner_annotations=tuple(learner_annotations),
+            )
+
+            # Map the authority_provided_id's of the course and all its sub-groupings to unified_course.
+            for authority_provided_id in course_authority_provided_ids:
+                self._unified_courses[authority_provided_id] = unified_course
+
+        return self._unified_courses
```

Comment:
Yikes, `unified_users()` and `unified_courses()` are massive methods that do loads of DB queries.

It's actually all the same queries that the code was doing before but now brought together into these two methods.

I could break up this code into private helper methods. But these two massive methods are the ones that I'm planning to rewrite in future PRs. The signatures, return values and tests hopefully won't need to change. But the method bodies should be completely replaced by ones that do a fixed number of larger queries.